### PR TITLE
Support success policy for TFJob

### PR DESCRIPTION
**Pull request by ChenYi015**
Cherry pick commits from https://github.com/kubeflow/training-operator/pull/1165.

### CODE REVIEW

**Comment by ChenYi015** [comment]
CC @cheyang @happy2048 @Syulin7

File: pkg/controller.v1/tensorflow/status.go [hunk 1]
```diff
@@ -103,8 +103,10 @@ func (tc *TFController) updateStatusSingle(tfjob *tfv1.TFJob, rtype tfv1.TFRepli
 		}
 	} else {
 		if rtype == tfv1.TFReplicaTypeWorker {
-			// All workers are succeeded or worker 0 completed, leave a succeeded condition.
-			if expected == 0 || worker0Completed {
+			// Leave a succeeded condition for the following two cases:
+			// 1. If default success policy is used and worker 0 has completed.
+			// 2. If `SuccessPolicyAllWorkers` success policy is used and all workers are succeeded.
+			if expected == 0 || (worker0Completed && *tfjob.Spec.SuccessPolicy != tfv1.SuccessPolicyAllWorkers) {
```
Comment (happy2048) [inline pkg/controller.v1/tensorflow/status.go]: should make sure tfjob.Spec.SuccessPolicy is not null

File: pkg/apis/tensorflow/v1/types.go [hunk 2]
```diff
@@ -56,6 +56,11 @@ type TFJobSpec struct {
 	// +optional
 	BackoffLimit *int32 `json:"backoffLimit,omitempty"`
 
+	// SuccessPolicy defines the policy to mark the TFJob as succeeded.
+	// Default to "", using the default rules.
+	// +optional
+	SuccessPolicy *SuccessPolicy `json:"successPolicy,omitempty"`
```
Comment (happy2048) [inline pkg/apis/tensorflow/v1/types.go:63]: can limit the successPolicy value by adding like: // +kubebuilder:validation:Enum=Lion;Wolf;Dragon?

File: pkg/controller.v1/tensorflow/status.go (hunk 1 above)
Comment (happy2048) [inline pkg/controller.v1/tensorflow/status.go]: conditions: 
1. successPolicy == AllWorkers and:
    a. all workers are succeeded 
    b. at least one worker is failed
2. successPolicy != AllWorkers and:
    a.worker0 is succeeded
    b.worker0 is failed

**Review by cheyang — APPROVED** [review]
/lgtm
/approve

File: pkg/apis/tensorflow/v1/common.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+// SuccessPolicy is the success policy.
+type SuccessPolicy string
+
+const (
+	SuccessPolicyChiefWorker SuccessPolicy = "ChiefWorker"
```
Comment (cheyang) [inline pkg/apis/tensorflow/v1/common.go]: I suggest keep consistency with Kubeflow training operator.

**Review by cheyang — CHANGES_REQUESTED** [review]

File: pkg/apis/tensorflow/v1/common.go (hunk 3 above)
Comment (ChenYi015) [reply pkg/apis/tensorflow/v1/common.go]: Done.

**Review by cheyang — APPROVED** [review]

**Pull request merged**